### PR TITLE
AP_AHRS/EKF: replace prearm_healthy() and prearm_failure_reason() with pre_arm_check()

### PR DESCRIPTION
**Pull request by rmackay9**
This PR enhances the AP_AHRS, DCM and EKF libraries' pre-arm check implementation consistent with other libraries.  This is a pre-requisite for my EKF position source PR: https://github.com/ArduPilot/ardupilot/pull/14803 which needs this change in order to cleanly handle pre-arm failures from multiple AHRS subsystems (i.e. EKF vs sensor source control library).  If someone wanted to add GSF specific pre-arm failure messages, this change should make that much easier.

The changes include:

- AP_AHRS::prearm_healthy() and AP_AHRS::prearm_failure_reason() methods are replaced with a single AP_AHRS::pre_arm_check() which is more consistent with prearm checks in other libraries.  It also removes the possibily of a mismatch between the results of the two calls.
- AP_AHRS_DCM implements the prearm_failure_reason and may return "AHRS not healthy" so there is no change in behaviour.  Previously DCM relied on the default implementation of prearm_healthy and did not implement the prearm_failure_reason.
- AP_AHRS_NavEKF logic is slighlty modified so it **could** (but rarely does) display one of these messages:
    - "AHRS: EKF2 not started"
    - "AHRS: EKF3 not started"
    - "AHRS: invalid EKF type" 
- EKF2 and EKF3 replace all_cores_healthy() and prearm_failure_reason() with pre_arm_check().  The change is the error message will be prefixed with "AHRS: " and these new messages may appear:
   - "AHRS: no EKFx cores"
   - "AHRS: EKFx core y unhealthy"

I don't mind removing the "AHRS: " prefix from EKF failures if people would prefer.

This has been lightly tested on Plane, Copter and Rover.
image

image

image

### CODE REVIEW

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 1]
```diff
@@ -1344,34 +1344,43 @@ bool AP_AHRS_NavEKF::healthy(void) const
...
+    const char *ahrs_prefix = "AHRS: ";
     switch (ekf_type()) {
 #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
     case EKFType::SITL:
 #endif
     case EKFType::NONE:
-        prearm_health = true;
-        break;
+        if (!healthy()) {
+            hal.util->snprintf(failure_msg, failure_msg_len, "AHRS not healthy");
```
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1357]: use ahrs_prefix

Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1357]:
```suggestion
            hal.util->snprintf(failure_msg, failure_msg_len, "Not healthy");
```

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 2]
```diff
@@ -1142,3 +1142,12 @@ bool AP_AHRS_DCM::get_velocity_NED(Vector3f &vec) const
     return true;
 }
 
+// returns false if we fail arming checks, in which case the buffer will be populated with a failure message
+bool AP_AHRS_DCM::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    if (!healthy()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "AHRS not healthy");
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:1149]:
```suggestion
        hal.util->snprintf(failure_msg, failure_msg_len, "Not healthy");
```

File: ArduCopter/AP_Arming.cpp [hunk 3]
```diff
@@ -507,12 +507,9 @@ bool AP_Arming_Copter::mandatory_gps_checks(bool display_failure)
...
     const AP_AHRS_NavEKF &ahrs = AP::ahrs_navekf();
-    if (!ahrs.prearm_healthy()) {
-        const char *reason = ahrs.prearm_failure_reason();
-        if (reason == nullptr) {
-            reason = "AHRS not healthy";
-        }
-        check_failed(display_failure, "%s", reason);
+    char failure_msg[50] = {};
+    if (!ahrs.pre_arm_check(failure_msg, sizeof(failure_msg))) {
+        check_failed(display_failure, "%s", failure_msg);
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:512]:
```suggestion
        check_failed(display_failure, "AHRS: %s", failure_msg);
```

File: ArduPlane/AP_Arming.cpp [hunk 4]
```diff
@@ -124,12 +124,9 @@ bool AP_Arming_Plane::ins_checks(bool display_failure)
...
         (checks_to_perform & ARMING_CHECK_INS)) {
-        if (!AP::ahrs().prearm_healthy()) {
-            const char *reason = AP::ahrs().prearm_failure_reason();
-            if (reason == nullptr) {
-                reason = "AHRS not healthy";
-            }
-            check_failed(ARMING_CHECK_INS, display_failure, "%s", reason);
+        char failure_msg[50] = {};
+        if (!AP::ahrs().pre_arm_check(failure_msg, sizeof(failure_msg))) {
+            check_failed(ARMING_CHECK_INS, display_failure, "%s", failure_msg);
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:129]:
```suggestion
            check_failed(ARMING_CHECK_INS, display_failure, "AHRS: %s", failure_msg);
```

File: ArduSub/AP_Arming_Sub.cpp [hunk 5]
```diff
@@ -65,12 +65,9 @@ bool AP_Arming_Sub::ins_checks(bool display_failure)
...
         (checks_to_perform & ARMING_CHECK_INS)) {
-        if (!AP::ahrs().prearm_healthy()) {
-            const char *reason = AP::ahrs().prearm_failure_reason();
-            if (reason == nullptr) {
-                reason = "AHRS not healthy";
-            }
-            check_failed(ARMING_CHECK_INS, display_failure, "%s", reason);
+        char failure_msg[50] = {};
+        if (!AP::ahrs().pre_arm_check(failure_msg, sizeof(failure_msg))) {
+            check_failed(ARMING_CHECK_INS, display_failure, "%s", failure_msg);
```
Comment (peterbarker) [inline ArduSub/AP_Arming_Sub.cpp:70]:
```suggestion
            check_failed(ARMING_CHECK_INS, display_failure, "AHRS: %s", failure_msg);
```

File: Rover/AP_Arming.cpp [hunk 6]
```diff
@@ -42,12 +42,9 @@ bool AP_Arming_Rover::gps_checks(bool display_failure)
...
     // always check if inertial nav has started and is ready
-    if (!ahrs.prearm_healthy()) {
-        const char *reason = ahrs.prearm_failure_reason();
-        if (reason == nullptr) {
-            reason = "AHRS not healthy";
-        }
-        check_failed(display_failure, "%s", reason);
+    char failure_msg[50] = {};
+    if (!ahrs.pre_arm_check(failure_msg, sizeof(failure_msg))) {
+        check_failed(display_failure, "%s", failure_msg);
```
Comment (peterbarker) [inline Rover/AP_Arming.cpp:47]:
```suggestion
        check_failed(display_failure, "AHRS: %s", failure_msg);
```

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 7]
```diff
@@ -1344,34 +1344,43 @@ bool AP_AHRS_NavEKF::healthy(void) const
     return AP_AHRS_DCM::healthy();
 }
 
-bool AP_AHRS_NavEKF::prearm_healthy(void) const
+// returns false if we fail arming checks, in which case the buffer will be populated with a failure message
+bool AP_AHRS_NavEKF::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
 {
-    bool prearm_health = false;
+    const char *ahrs_prefix = "AHRS: ";
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1350]:
```suggestion
```

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 8]
```diff
@@ -1344,34 +1344,43 @@ bool AP_AHRS_NavEKF::healthy(void) const
...
+            return false;
+        }
+        return true;
 
 #if HAL_NAVEKF2_AVAILABLE
     case EKFType::TWO:
-        if (_ekf2_started && EKF2.all_cores_healthy()) {
-            prearm_health = true;
+        if (!_ekf2_started) {
+            hal.util->snprintf(failure_msg, failure_msg_len, "%sEKF2 not started", ahrs_prefix);
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1365]:
```suggestion
            hal.util->snprintf(failure_msg, failure_msg_len, "EKF2 not started");
```

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 9]
```diff
@@ -1344,34 +1344,43 @@ bool AP_AHRS_NavEKF::healthy(void) const
...
-        break;
+        return EKF2.pre_arm_check(failure_msg, failure_msg_len);
 #endif
 
 #if HAL_NAVEKF3_AVAILABLE
     case EKFType::THREE:
-        if (_ekf3_started && EKF3.all_cores_healthy()) {
-            prearm_health = true;
+        if (!_ekf3_started) {
+            hal.util->snprintf(failure_msg, failure_msg_len, "%sEKF3 not started", ahrs_prefix);
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1374]:
```suggestion
            hal.util->snprintf(failure_msg, failure_msg_len, "EKF3 not started");
```

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 10]
```diff
@@ -1344,34 +1344,43 @@ bool AP_AHRS_NavEKF::healthy(void) const
...
+            return false;
         }
-        break;
+        return EKF3.pre_arm_check(failure_msg, failure_msg_len);
 #endif
     }
-   return prearm_health && healthy();
+
+    // if we get here then ekf type is invalid
+    hal.util->snprintf(failure_msg, failure_msg_len, "%sinvalid EKF type", ahrs_prefix);
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:1382]:
```suggestion
    hal.util->snprintf(failure_msg, failure_msg_len, "invalid EKF type");
```

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp [hunk 11]
```diff
@@ -906,13 +906,22 @@ bool NavEKF2::healthy(void) const
...
+    const char *ahrs_prefix = "AHRS: ";
     if (!core) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "%sno EKF2 cores", ahrs_prefix);
         return false;
     }
     for (uint8_t i = 0; i < num_cores; i++) {
         if (!core[i].healthy()) {
+            const char *failure = core[primary].prearm_failure_reason();
+            if (failure != nullptr) {
+                hal.util->snprintf(failure_msg, failure_msg_len, "%s%s", ahrs_prefix, failure);
```
Comment (peterbarker) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp:921]:
```suggestion
                hal.util->snprintf(failure_msg, failure_msg_len, "%s", failure);
```

**Review by tridge — DISMISSED** [review]

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp [hunk 12]
```diff
@@ -906,13 +906,22 @@ bool NavEKF2::healthy(void) const
...
+            } else {
+                hal.util->snprintf(failure_msg, failure_msg_len, "%sEKF2 core %d unhealthy", ahrs_prefix, (int)i);
```
Comment (peterbarker) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp:923]:
```suggestion
                hal.util->snprintf(failure_msg, failure_msg_len, "EKF2 core %d unhealthy", (int)i);
```

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 13]
```diff
@@ -1030,13 +1030,22 @@ bool NavEKF3::healthy(void) const
     return core[primary].healthy();
 }
 
-bool NavEKF3::all_cores_healthy(void) const
+// returns false if we fail arming checks, in which case the buffer will be populated with a failure message
+bool NavEKF3::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
 {
+    const char *ahrs_prefix = "AHRS: ";
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:1036]:
```suggestion
```

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 14]
```diff
@@ -1030,13 +1030,22 @@ bool NavEKF3::healthy(void) const
...
     if (!core) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "%sno EKF3 cores", ahrs_prefix);
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:1038]:
```suggestion
        hal.util->snprintf(failure_msg, failure_msg_len, "no EKF3 cores");
```

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 15]
```diff
@@ -1030,13 +1030,22 @@ bool NavEKF3::healthy(void) const
...
         return false;
     }
     for (uint8_t i = 0; i < num_cores; i++) {
         if (!core[i].healthy()) {
+            const char *failure = core[primary].prearm_failure_reason();
+            if (failure != nullptr) {
+                hal.util->snprintf(failure_msg, failure_msg_len, "%s%s", ahrs_prefix, failure);
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:1045]:
```suggestion
                hal.util->snprintf(failure_msg, failure_msg_len, "%s", failure);
```

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 16]
```diff
@@ -1030,13 +1030,22 @@ bool NavEKF3::healthy(void) const
...
+            } else {
+                hal.util->snprintf(failure_msg, failure_msg_len, "%sEKF3 core %d unhealthy", ahrs_prefix, (int)i);
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:1047]:
```suggestion
                hal.util->snprintf(failure_msg, failure_msg_len, "EKF3 core %d unhealthy", (int)i);
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: I've included all the changes suggested by @peterbarker

**Comment by rmackay9** [comment]
I managed to reproduce the following AHRS messages:

AHRS: EKF3 not started
AHRS: EKF2 not started
PreArm: AHRS: GPS speed error 1.6 (needs < 1.0)
PreArm: AHRS: EKF2 core 1 unhealthy

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: Txs for the review.  AHRS prefix added so this message now appears as, "AHRS: Not healthy"

**Review by peterbarker — APPROVED** [review]

**Comment by peterbarker** [comment]
travis has actually passed -  results haven't been passed back.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**